### ISSUE TITLE
BUG: Unexpected interpretation of explicit list

### ISSUE THREAD

**Issue by jeswr** [eyereasoner/eye#73]
Unexpected interpretation of explicit list

`data.n3`
```n3

@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
[] rdf:first 0 ;
  rdf:rest rdf:nil .
```

Is interpreted as

```
$ npx eyereasoner --nope --quiet --pass-all ./listTest.n3 
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.

(0) rdf:first 0 .
(0) rdf:rest rdf:nil.
```

**Comment by jeswr** [comment]
Also if I have 

```
{} => {
  [] rdf:first 0 ;
  rdf:rest rdf:nil .
} .
```

Then there is no output

**Comment by jeswr** [comment]
```
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix : <https://example.org/ns/1/> .

(0 1) :p :o, :o2 .
```

Also incorrectly becomes

```
$ npx eyereasoner --nope --quiet --pass-all ./listTest.n3 
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix : <https://example.org/ns/1/>.

(0 1) :p :o.
(0 1) :p :o2.
```

**Comment by jeswr** [comment]
```
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix : <https://example.org/ns/1/> .

:a :b [
  rdf:first 0 ;
  rdf:rest rdf:nil ;
  :p :o ;
] .
```

Incorrectly becomes

```
$ npx eyereasoner --nope --quiet --pass-all ./listTest.n3 
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
@prefix : <https://example.org/ns/1/>.

:a :b [ rdf:first 0; rdf:rest rdf:nil].
_:bn_1 :p :o.
```

**Comment by jeswr** [comment]
Also note that the following derives `:a :a 2`

```
@prefix math: <http://www.w3.org/2000/10/swap/math#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix : <http://example.org/> .

{
  (3 1) math:difference ?o .
} => {
  :a :a ?o .
} .
```

Whilst the following does not (but should)

```
@prefix math: <http://www.w3.org/2000/10/swap/math#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix : <http://example.org/> .

{
  [] math:difference ?o ;
     rdf:first 3 ;
     rdf:rest [ rdf:first 1 ;
                rdf:rest rdf:nil ] .
} => {
  :a :a ?o .
} .
```

**Comment by josd** [comment]
I tried hard with the last one but it is impossible to use rdf:first rdf:rest ladders as subject or object of built-ins because they heavily interfere with coroutining (suspending the built-in till for instance particular arguments are bound).

**Comment by josd** [comment]
> Unexpected interpretation of explicit list
> 
> `data.n3`
> 
> ```
> 
> @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
> [] rdf:first 0 ;
>   rdf:rest rdf:nil .
> ```
> 
> Is interpreted as
> 
> ```
> $ npx eyereasoner --nope --quiet --pass-all ./listTest.n3 
> @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
> 
> (0) rdf:first 0 .
> (0) rdf:rest rdf:nil.
> ```

Might seem unexpected but is not broken.

**Comment by josd** [comment]
> Unexpected interpretation of explicit list
> 
> `data.n3`
> 
> ```
> 
> @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
> [] rdf:first 0 ;
>   rdf:rest rdf:nil .
> ```
> 
> Is interpreted as
> 
> ```
> $ npx eyereasoner --nope --quiet --pass-all ./listTest.n3 
> @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
> 
> (0) rdf:first 0 .
> (0) rdf:rest rdf:nil.
> ```

EYE v3.3.1 now gives
```
(0) rdf:first 0 .
(0) rdf:rest ().
```
which might look unexpected but nothing is broken.

**Comment by josd** [comment]
> Also if I have
> 
> ```
> {} => {
>   [] rdf:first 0 ;
>   rdf:rest rdf:nil .
> } .
> ```
> 
> Then there is no output

EYE v3.3.1 now gives
```
_:sk_0 rdf:first 0 .
_:sk_0 rdf:rest ().
```

**Comment by josd** [comment]
> ```
> @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
> @prefix : <https://example.org/ns/1/> .
> 
> (0 1) :p :o, :o2 .
> ```
> 
> Also incorrectly becomes
> 
> ```
> $ npx eyereasoner --nope --quiet --pass-all ./listTest.n3 
> @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
> @prefix : <https://example.org/ns/1/>.
> 
> (0 1) :p :o.
> (0 1) :p :o2.
> ```

Nothing is broken here.

**Comment by josd** [comment]
> ```
> @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
> @prefix : <https://example.org/ns/1/> .
> 
> :a :b [
>   rdf:first 0 ;
>   rdf:rest rdf:nil ;
>   :p :o ;
> ] .
> ```
> 
> Incorrectly becomes
> 
> ```
> $ npx eyereasoner --nope --quiet --pass-all ./listTest.n3 
> @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
> @prefix : <https://example.org/ns/1/>.
> 
> :a :b [ rdf:first 0; rdf:rest rdf:nil].
> _:bn_1 :p :o.
> ```
EYE v3.3.1 now gives
```
:a :b (0).
_:bn_1 :p :o.
```

**Comment by josd** [comment]
> > ```
> > @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
> > @prefix : <https://example.org/ns/1/> .
> > 
> > :a :b [
> >   rdf:first 0 ;
> >   rdf:rest rdf:nil ;
> >   :p :o ;
> > ] .
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Incorrectly becomes
> > ```
> > $ npx eyereasoner --nope --quiet --pass-all ./listTest.n3 
> > @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>.
> > @prefix : <https://example.org/ns/1/>.
> > 
> > :a :b [ rdf:first 0; rdf:rest rdf:nil].
> > _:bn_1 :p :o.
> > ```
> 
> EYE v3.3.1 now gives
> 
> ```
> :a :b (0).
> _:bn_1 :p :o.
> ```

Corrected in EYE v3.3.3 and now giving
```
:a :b (0).
(0) :p :o.
```
